### ISSUE TITLE
Update projects to build net7.0 TFM

### ISSUE THREAD

**Issue by MichaelSimons** [dotnet/roslyn#62510]
Source build requires projects to build with the current TFM and/or netstandard TFMs. roslyn is targeting net6.0 and netstandard2.0. Ideally the net6.0 reference would be updated to net7.0. If that is not possible, the TFM could be conditioned for source-build but ideally the regular and source build would be identical.

**Comment by RikkiGibson** [comment]
Could you help us understand that requirement better? I think Roslyn tends to target the most recent LTS which has been released to production. For example, #57567 was merged the day net6.0 came out.

I wasn't sure if we were actually expecting to target net7.0 or to simply wait until the next LTS to update the target (net8.0?). cc @jaredpar

**Comment by jaredpar** [comment]
> Ideally the net6.0 reference would be updated to net7.0. 

I do not believe we are in a situation where we can change `net6.0` to `net7.0`. Consider that the C# IDE team references our binaries and they must run on a `net6.0` runtime (that is the private runtime that executes inside of Visual Studio). Flipping `net6.0` to `net7.0` would cause the C# IDE to suddenly use our `nestandard2.0` binaries which is likely unexpected. Essentially there are several requirements on us here and one of them is holding us to `net6.0` for the foreseeable future. 

It's possible we could add `net7.0` to our `<TargetFrameworks>` set for the core projects that intersect source build. But that is going to add build time and force us to think about our test matrix a bit. It either creates a test gap for us or forces us to significantly increase our test matrix. 

> Source build requires projects to build with the current TFM and/or netstandard TFMs

Roslyn has essentially never been on latest TFM for basically the reasons I outlined above. How has this been handled in the past?

**Comment by jaredpar** [comment]
@vatsalyaagrawal 

> It's possible we could add net7.0 to our <TargetFrameworks> set for the core projects that intersect source build. But that is going to add build time and force us to think about our test matrix a bit. It either creates a test gap for us or forces us to significantly increase our test matrix.

This is another case where putting all of our eggs into the PR basket is hurting us a bit. From a code perspective testing both `net6.0` and `net7.0` isn't that much work. Need to update a number of compiler test projects to include `net7.0` in the multi-target list and probably a few changes to `eng/build.ps1`. That will result in a **significant** increase in the number of tests we run on every PR. 

This is a case where the tests are necessary, compiler frequently discovers important behavior changes between runtimes. At the same time the number of instances where that happens is incredibly low. Probably find a gap 2-4 times a year. 

If we had a strong monitoring of our CI runs then I'd feel comfortable with; 

- Including `net7.0` in our `<TargetFrameworks>` list 
- Only running `net7.0` on CI and excluding it from PR 

That would keep the hit to the PR time to be pretty minimal. Just a handful of more compilation events. The `net7.0` failures could be dealt with in an async fashion. 

@jmarolf FYI as I know you've been pushing on the PR times so wanted you to be aware of this discussion on trade offs

**Comment by crummel** [comment]
That testing approach makes sense to me.  I don't think it's sufficient for source-build though to add the net7.0 TFMs; the problem is more the presence of TFMs that we aren't building for since we can't produce the 6.0 targeting pack in the 7.0 build - source-build would still have to flip a switch to only produce net7.0 bits.

**Comment by jaredpar** [comment]
> This would lead me to believe that something like this is the right approac

That is roughly the pattern I have in mind. The one issue I have remaining is who is responsible for monitoring our CI test failures? Basically when this does fail, and it will from time to time, who is on the hook for monitoring it and firing off an email / issue / etc ... within 1/2/3 days?
